model: EmbeddingGemma Adding Support for SentenceTransformers Dense Modules

sfallah opened the pull request:
### model: add support for EmbeddingGemma SentenceTransformers dense linear projections

- Adding support for the Dense modules used in EmbeddingGemma models.

- EmbeddingGemma is a SentenceTransformers model with additional modules beyond the base Transformer backbone.

- See: https://developers.googleblog.com/en/gemma-explained-embeddinggemma-architecture-and-recipe/

sfallah committed:
model: EmbeddingGemma sentence-transformers dense linear projections support

sfallah committed:
model: add support for EmbeddingGemma SentenceTransformers dense linear projections

Adding support for the Dense modules used in EmbeddingGemma models.
EmbeddingGemma is a SentenceTransformers model with additional modules beyond the base Transformer backbone.

See: https://developers.googleblog.com/en/gemma-explained-embeddinggemma-architecture-and-recipe/

sfallah commented:
@danbev
Thanks for your rapid feedback.

I had already anticipated the reasons why dense layers were not included in the first place.
And I understand your arguments for not including the additional SentenceTransformers (ST) modules — but in practice, for the client/user, it’s very helpful if they are already included.

Let’s take the example of my own project, where I’m using `embeddinggemma` for RAG.
First and foremost, I would face a quality issue if dense layers are not applied.
Besides the fact that, as far as I know, the MTEB benchmarks for `embeddinggemma` are done using the ST model (not just the base model), the quality issue can even be demonstrated with a trivial test.

Please see the example below:

**Base Model:**
```
A man is playing guitar:
	The dog plays in the garden: 0.5552932620048523
	A woman watches TV: 0.4669498801231384
	Do you like pizza?: 0.4650737941265106
	The new movie is so great: 0.438459575176239
I love pasta:
	Do you like pizza?: 0.676074206829071
	The new movie is so great: 0.5771214962005615
	A woman watches TV: 0.4970381259918213
	The dog plays in the garden: 0.4736071228981018
```
	
	
**ST Model:**
```
A man is playing guitar:
	The dog plays in the garden: 0.5132929682731628
	A woman watches TV: 0.43405574560165405
	The new movie is so great: 0.3694726824760437
	Do you like pizza?: 0.31990543007850647
I love pasta:
	Do you like pizza?: 0.6048902273178101
	A woman watches TV: 0.38842126727104187
	The new movie is so great: 0.3778345584869385
	The dog plays in the garden: 0.33146393299102783
```

The results become even more different (not to say worse) when MRL-reduced dimensions are used.
So for me as a user — someone who wants to use this for RAG and similar applications — only the full ST model is truly useful.
And if I were to apply the dense layers myself on the client side, it would be quite impractical and most likely inefficient.

As a user, I would have preferred the following:

* That `llama.cpp` supports both base-model-only and ST-model modes.
* That `convert` supports either full ST conversion or base-model-only conversion.
* That both ST GGUF models and base models can be loaded.
* (Optional) That applying the dense layers can be toggled in server embedding requests.

I could also imagine accommodating or implementing ST modules in a more generic way, similar to how LoRA adapters are handled.

Sorry for making this so long, but this model is an important one for users like me.
It’s very efficient and has a high MTEB ranking — but for that to hold true, the dense layers are crucial.

ggerganov commented:
@sfallah Thanks for the detailed description - this is quite helpful. The main reason to not have support for Dense embedding modules implemented is that until recently (i.e. until our work with @danbev on EmbeddingGemma) I had no idea what their purpose is and how they are used. But now it is more clear.

We should add some way to support that. It seems it would involve generalizing/extending the pooling logic/API as well as (optionally) incorporating the modules (i.e. tensors) into the GGUFs during conversion.

> (Optional) That applying the dense layers can be toggled in server embedding requests.

On first thought, the configuration of the dense modules would have to be done on the `llama_context` level, so dynamically switching the modules per request might not be possible to support. But if this is an important use case, we can think of ways to accommodate it, thought it would require changes in both the server embedding API and the `libllama` API 🤔

Since you have some first steps towards adding support for dense modules with this PR, we can continue with designing and implementing support for dense module configurations. Let me know if you are interested in putting some extra work into this, and I will try to provide steps how to proceed.

sfallah commented:
@ggerganov
Thank you for taking the time to review this PR.
I’m very interested in collaborating with you and the team on this.

sfallah commented:
@ggerganov
I totally agree, if I may say, this is basically what I also meant with my suggestions.

In my opinion, it is not a burden for the user—at least not for me—to know if I will be deploying a gguf-model that includes dense layers or not. The same way, as I need to know which quantization type my gguf-model has.

The only issue is flexibility regarding pooling, which I would, in practice, not see as a problem because of the following reasons:

1. What pooling type is "practically" applicable to what model is model-dependent.
2. The option is practically binary: between the model's "default" pooling or none.

I know the second point is a bit oversimplified, but in practice, it is generally true for `embedding` and `reranker` models.

So I don't see any problem if, for example, in the case of the `embeddinggemma` model with dense layers included, the pooling set by the user is ignored, because dense layers require mean pooling.

sfallah committed:
model: add support for EmbeddingGemma SentenceTransformers dense linear projections

- converting model with dense-layers is optional
- introduced dense config params

sfallah committed:
Merge branch 'master' into embeddinggemma_sentence_transformers

danbev commented on src/llama-hparams.h:
```diff
@@ -169,6 +169,15 @@ struct llama_hparams {
     uint32_t n_embd_altup = 256;
 
+    // needed for sentence-transformers dense layers
+    uint32_t dense_2_feat_in = 0; // in_features of the 2_Dense
```
Nit: Can you lines these up like the other fields in this struct.

danbev replied:
@sfallah Sorry I was not clear about what I meant. I mean like in https://github.com/ggml-org/llama.cpp/blob/c5fef0fcea3b978a2318b1af170209ecec7c37b4/src/llama-hparams.h#L166-L170 where the fields lineup. Like this:
```c++
    // needed for sentence-transformers dense layers
    uint32_t dense_2_feat_in  = 0; // in_features of the 2_Dense
    uint32_t dense_2_feat_out = 0; // out_features of the 2_Dense
    uint32_t dense_3_feat_in  = 0; // in_features of the 3_Dense
    uint32_t dense_3_feat_out = 0; // out_features of the 3_Dense
```

iamlemec commented:
Very cool! My main question here is: why not just add another pooling type `DENSE`? For the `RANK` case, we are already pulling in additional weight matrices for the classification head (post token pooling). Seems like this adds a bunch of flags and struct members that ar e very specific to EmbeddingGemma.

In fact, you could even use the existing tensor placeholders `cls` and `cls_out` to store the first and second dense layers. Then the code in `build_pooling` would be nearly identical to the RANK case (minus the `tanh`). As for the conversion, I would think you could just make some modifications to `TextModel._try_set_pooling_type` to detect the presence of dense layers. That looks through the `modules.json` file for more info.

sfallah committed:
Update convert_hf_to_gguf.py

Co-authored-by: Daniel Bevenius <daniel.bevenius@gmail.com>

sfallah committed:
fixed formatting issues

ggerganov reviewed (commented):
Overall looking good.

From what I understand, we currently assume that the output features from the dense modules would be equal to `hparams.n_embd`. I guess this is usually true, but maybe in the future we would have to add support for cases where it's not true. This would require to add some way for the user to query the number of output features through the `libllama` API.

Regarding @iamlemec proposal about adding a new `DENSE` pooling type - I think a better alternative would be to separate the `RANK` and `DENSE` concepts away from the pooling type. For example, reranking assumes a certain pooling type - currently `LAST` from what we've seen. But there is no guarantee that in the future some model would not use different pooling type to do reranking. Same argument is valid for the dense modules. So the 2 concepts would have to be separated. For now, this PR is OK - we can work on this after.

sfallah committed:
Update src/llama-graph.cpp

Co-authored-by: Georgi Gerganov <ggerganov@gmail.com>

danbev commented:
I've updated https://github.com/ggml-org/llama.cpp/pull/16387 to support the `--sentence-transformers-dense-modules` flag.

sfallah committed:
- removed pooling_type_opt, always allow overriding pooling_type
- asserts checking dense features dims

sfallah committed:
fix python lint

sfallah committed:
fix ubuntu gcc build warning

sfallah committed:
Merge branch 'ggml-org:master' into embeddinggemma_sentence_transformers

sfallah committed:
Merge branch 'ggml-org:master' into embeddinggemma_sentence_transformers

sfallah committed:
- fixed thread-safety test
- moved asserts to load_hparams

ggerganov commented:
> This will require us to publish separate models for these "ST models" if I've understood this correctly?

I don't think it's necessary to publish the ST models. For now this will be an option for advanced users as explained in https://github.com/ggml-org/llama.cpp/pull/16367#issuecomment-3360551675. I think that soon we will extend the API as explained in https://github.com/ggml-org/llama.cpp/pull/16367#pullrequestreview-3308508283 to support specifying the "post-pooling" parameters and this will allow to have a single GGUF that can be used both for raw pooled embedding and for ST embeddings based on the `llama_context` configuration.

sfallah committed:
- tidying up code
- simplifying graph-context expecting both dense weights

ggerganov committed:
minor : add TODO [no ci]
